src/inventory/transaction/EnchantTransaction.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\inventory\transaction;
+
+use pocketmine\block\inventory\EnchantInventory;
+use pocketmine\inventory\transaction\action\InventoryAction;
+use pocketmine\item\Item;
+use pocketmine\item\ItemTypeIds;
+use pocketmine\player\Player;
+use function count;
+
+class EnchantTransaction extends InventoryTransaction{
+
+	private int $optionId;
+	private ?Item $inputItem = null;
+	private int $lapisCost = 0;
+	private int $enchantmentLevel = 0;
+
+	public function __construct(Player $source, int $optionId){
+		$this->optionId = $optionId;
+		parent::__construct($source);
+	}
+
+	public function validate() : void{
+		if(count($this->actions) < 1){
+			throw new TransactionValidationException("Transaction must have at least one action to be executable");
+		}
+
+		$enchantWindow = $this->source->getCurrentWindow();
+		if(!$enchantWindow instanceof EnchantInventory){
+			throw new TransactionValidationException("Current window was expected to be of type EnchantInventory");
+		}
+
+		$enchantmentLevel = $enchantWindow->getOptionEnchantmentLevel($this->optionId);
+		if($enchantmentLevel === null){
+			throw new TransactionValidationException("Incorrect option id $this->optionId");
+		}
+		$this->enchantmentLevel = $enchantmentLevel;
+
+		if($this->inputItem === null || !$this->inputItem->equalsExact($enchantWindow->getInput())){
+			throw new TransactionValidationException("Incorrect input item");
+		}
+
+		if($this->source->hasFiniteResources()){
+			if($this->lapisCost !== $enchantmentLevel){
+				throw new TransactionValidationException("Expected the amount of lapis lazuli spent to be $enchantmentLevel, but received $this->lapisCost");
+			}
+
+			$xpLevel = $this->source->getXpManager()->getXpLevel();
+			if($xpLevel < $enchantmentLevel){
+				throw new TransactionValidationException("Expected player to have xp level at least of $enchantmentLevel, but received $xpLevel");
+			}
+		}
```

Comment:
The validation here is implemented incorrectly.

Typically, transaction balance should be calculated, which gives us a list of destroyed items (the inputs) and a list of created items (the outputs). This then tells the amount of lapis used and the item consumed as input. The destroyed input should then be enchanted with the enchantments selected and compared to the output. If the output doesn't match, the transaction is invalid.

The purpose of this is to allow plugins to create and validate such transactions using only slot changes.

You can look at CraftingTransaction to see how this should be done.

Reply:
If we need the ability to enchant without an enchant inventory, then it turns out we need to store the enchantment options in some other class than EnchantInventory?

Reply:
And `NetworkSession::sendEnchantmentOptions` will refer to this class (say EnchantmentOptionManager) to save the options?